### ISSUE TITLE
Will Tensorflow 2.0 support synchronizing batchnorm statistics across multiple GPUs?

### ISSUE THREAD

**Issue by panfengli** [tensorflow/tensorflow#22755]
Have I written custom code: N/A
OS Platform and Distribution: cpe:/o:centos:centos:7
TensorFlow installed from: pip install 
TensorFlow version: 1.10
Bazel version: N/A
CUDA/cuDNN version: Cuda 9.0
GPU model and memory: 1080 Ti, 11G
Exact command to reproduce: N/A
Mobile device: N/A

I checked the [roadmap](https://www.tensorflow.org/community/roadmap) says the new feature:

```
Eager execution

    Use DistributionStrategy to utilize multiple GPUs and multiple TPU cores.
    Distributed training support (multi-machine).
    Performance improvements.
    Simpler export to a GraphDef/SavedModel.
```

I am wondering whether tf 2.0 will consider synchronizing batchnorm statistics across multiple GPUs which is extremely helpful for segmentation and other task.

Similar issue as [Synchronized BatchNorm statistics across GPUs](https://github.com/tensorflow/tensorflow/issues/18222)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by panfengli** [comment]
A implementation in pytorch for reference:

 [Implementing Synchronized Multi-GPU Batch Normalization](https://hangzhang.org/PyTorch-Encoding/notes/syncbn.html)

Code in cuda:
[syncbn_kernel.cu](https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/lib/gpu/syncbn_kernel.cu)

**Comment by panfengli** [comment]
MxNet also provided a version for sync bc layer:

[mxnet.gluon.contrib.nn.SyncBatchNorm](http://mxnet.incubator.apache.org/versions/master/api/python/gluon/contrib.html?highlight=syncbatchnorm#mxnet.gluon.contrib.nn.SyncBatchNorm)

**Comment by ppwwyyxx** [comment]
There is an open issue about this (as you have linked to in your post), with some working solutions already. So it's better to leave this issue closed and marked as duplication, to make the discussion more focused.

**Comment by panfengli** [comment]
@ppwwyyxx Yeah, I see you have implemented in tensorpack. But I wish it could at least provided in `tf.contrib`, since generally the original framework has better maintenance especially when new version comes out.

Since Tensorflow 2.0 claims that it will support "Use DistributionStrategy to utilize multiple GPUs and multiple TPU cores", and `tf.contrib` will be deprecated, so that I am wondering whether sync bn will be provided in TF 2.0.

I don't think it is a duplication, at least currently sync bn is not provided in Tensorflow. As I mentioned, there do exist a lot of working solutions, the one in pytorch, in tensorpack, but they are not in official Tensorflow or Pytorch package, once Tensorflow or Pytorch upgrades to newer and newer version, there might be disintegration.

**Comment by ppwwyyxx** [comment]
I agree it's best if the feature exists in tensorflow. And that's actually exactly what the existing issue is about and the reason why the issue remains open.

**Comment by byronyi** [comment]
Let’s just close this issue. It’s duplicated.

To be more constructive, please submitting PR to support new features, not duplicated issues.

**Comment by seemuch** [comment]
Please follow #18222 since this is duplicate.

**Comment by jianlong-yuan** [comment]
I just implement this function [https://github.com/jianlong-yuan/syncbn-tensorflow](url)

**Comment by panfengli** [comment]
@jianlong-yuan Thanks for your work! Does it follow the similar way in cifar multi gpu example, especially when combining different towers' gradient results like:

```
def average_gradients(tower_grads):
    """Calculate the average gradient for each shared variable across all towers.
      Note that this function provides a synchronization point across all towers.

    Args:
        tower_grads: List of lists of (gradient, variable) tuples. The outer list
        is over individual gradients. The inner list is over the gradient
        calculation for each tower.
    Returns:
        List of pairs of (gradient, variable) where the gradient has been averaged
        across all towers.
    """
    average_grads = []
    for grads_and_vars in zip(*tower_grads):
        # Each grad_and_vars looks like the following:
        # ((grad0_gpu0, var0_gpu0), ... , (grad0_gpuN, var0_gpuN))
        grads = [g for g,_ in grads_and_vars]
        if grads[0] is not None: # Gradient can be None
            # Average the grads
            grad = tf.reduce_mean(tf.stack(grads, 0), 0)
            # The Variables are redundant because they are shared across towers
            v = grads_and_vars[0][1]
            grad_and_var = (grad, v)
        else:
            grad_and_var = grads_and_vars[0]

        average_grads += [grad_and_var]
    return average_grads
```

**Comment by byronyi** [comment]
FYI, TF just added Sync BN: adf769043f0c48a44c05c5a24aac14f0b4951896
